### DISCUSSION TITLE
Changing tick-values in raster-plot

### DISCUSSION THREAD

**Discussion by Hvass-Labs** [observablehq/plot#1987] [Q&A] [answered]
Consider this [gallery example](https://observablehq.com/@observablehq/plot-volcano-raster?intent=fork). How can I change x-axis tick values to go from e.g. 700 to 1000 instead of from 0 to 87, which is just an enumeration of the input data?

I've tried changing the domain, range and ticks args, but I can't get it working.

Thanks!

**Comment by Fil** [comment, marked as answer]
If you want to keep *data* as an array of pixel colors, the best solution is to apply an affine transform on x and y:

```
Plot.plot({
  aspectRatio: 1,
  color: {label: "Elevation (m)", legend: true},
  x: {transform: d => d * 10 - 100},
  y: {transform: d => d * 10 + 223},
  marks: [
    Plot.raster(volcano.values, {width: volcano.width, height: volcano.height})
  ]
})
```
![transform-x-y](https://github.com/observablehq/plot/assets/7001/9678aad5-78d4-4efb-8935-363b4235b6b2)

Otherwise you can see data as an array of **samples** with abstract positions and values, and use a spatial interpolator to fill the image; but it's not going to give the same image (for better or worse).

```
Plot.plot({
  aspectRatio: 1,
  color: {label: "Elevation (m)", legend: true},
  marks: [
    Plot.raster(volcano.values, {
      fill: Plot.identity,  // the value is the color
      x: (d, i) => (i % volcano.width) * 10 - 100, 
      y: (d, i) => Math.floor(i / volcano.width) * 10 + 223,
      interpolate: "random-walk" 
    })
  ]
})
```

![sample-interpolate](https://github.com/observablehq/plot/assets/7001/155d24d5-ef06-4b1c-8046-f8763a6e173d)

**Reply by Hvass-Labs** [reply]
Thanks very much!

The first solution also works for my actual problem, but I was having trouble getting the second to work. Furthermore, I need to add more marks to the plot that use the "correct" domains for the data-values, and the first solution just changes the displayed tick-values.

I wonder if it would be better if I convert the simple 1-dim array to e.g. an array of dicts with proper x/y coordinates and values like this:

    volcano2 = [{x: 123, y: 456, value: 78}, {x: 124, y: 456, value: 23}, ...]

And then the main plotting would be more like this:

    Plot.raster(volcano2, {x: "x", y: "y", fill: "value"})

Would you recommend this method instead? And do you already have some clever functions for doing such array conversion? Thanks!

**Reply by Fil** [reply]
You can use the functions I gave above to map x and y:

```
x: (d, i) => (i % volcano.width)
y: (d, i) => Math.floor(i / volcano.width)
```

this gives you the grid coordinates:
```
x=0, 1,2, 3, 4… w-1, 0, 1, 2, 3… w-1 etc
y=0, 0, 0, 0, 0… 0, 1, 1, 1, 1… 1 etc.
```

then use any affine transform on this `(x, y) => [a x+b, a y + c];` to scale and translate?

The only "clever" trick I'd suggest is to compute X and Y as arrays of numbers, and pass them as channels directly to the mark. It will be faster.

**Reply by Hvass-Labs** [reply]
Thanks. Please elaborate what you mean by _"... compute X and Y as arrays of numbers, and pass them as channels directly to the mark. It will be faster."_

I thought that Plot only accepted arrays of dicts like this:

    data1 = [{x: 1, y: 200}, {x: 2, y: 300}, ...]

Are you saying that it also accepts data like this?

    data2 = {x: [1, 2, ...], y: [200, 300, ...]}

That would make my life a lot easier, as I have had to make all kinds of conversion functions to get my data in the `data1` format above. It would have been much easier for me to use the `data2` format. But I just tried it with the following simple example in a HQ Notebook, and it just produces an empty plot?

    data = ({x: aapl.map(d => d.Date), y: aapl.map(d => d.Close)});
    Plot.lineY(data, {x: "x", y: "y"}).plot({y: {grid: true}})

**Reply by Fil** [reply]
see the documentation for channels: https://observablehq.com/plot/features/marks#marks-have-channels

**Reply by Hvass-Labs** [reply]
This doesn't seemt to be documented under ["Marks have channels"](https://observablehq.com/plot/features/marks#marks-have-channels), but the following is mentioned under ["Mark options"](https://observablehq.com/plot/features/marks#mark-options)

> Plot also supports columnar data for greater efficiency with bigger datasets; for example, data can be specified as any array of the appropriate length (or any iterable or value compatible with Array.from), and then separate arrays of values can be passed as options.

My small example from above works like this:

    X = aapl.map(d => d.Date);
    Y = aapl.map(d => d.Close);
    index = [...Array(X.length).keys()];
    Plot.lineY(index, {x: X, y: Y}).plot({y: {grid: true}})

This is a complete surprise to me. I think this needs to be documented more clearly, so I'll open another issue, as this issue is getting derailed.

**Reply by Fil** [reply]
new topic is https://github.com/observablehq/plot/discussions/1989

**Reply by Hvass-Labs** [reply]
You answered my original question so I'm closing this. I still don't quite understand how to use array channels, so I've opened #1990 with that question. Thanks!
